File: gh360_examples/gh360_examples/camera_frame.py
from geometry_msgs.msg import TransformStamped, Transform, Vector3, Quaternion

import rclpy
from rclpy.node import Node
import tf2_ros
from tf2_ros import TransformBroadcaster
from scipy.spatial.transform import Rotation


class CameraFrameBroadcaster(Node):
    def __init__(self):
        super().__init__('camera_frame_tf2_broadcaster')
        self.tf_broadcaster = TransformBroadcaster(self)

        rot = []
        rot.append([0.42261826174069345,	0.38302221852676516,	0.8213938062574561])
        rot.append([7.415938757451997e-15,	0.9063077885970263,	-0.42261825839446004])
        rot.append([-0.9063077885970263,	0.1802399555017374,	0.3830222218690612])
        sci_rot = Rotation.from_matrix(rot)
        sci_quat = sci_rot.as_quat()
        quat = Quaternion(x=sci_quat[0], y=sci_quat[1], z=sci_quat[2], w=sci_quat[3])
        
        trans = Vector3(x=0.13395628663621322, y=0.07813541504103579, z=0.06372504768866594)

        self.transform  = Transform(rotation=quat,translation=trans)

        self.timer = self.create_timer(0.1, self.broadcast_timer_callback)



    def broadcast_timer_callback(self):
        t = TransformStamped()

        t.header.stamp = self.get_clock().now().to_msg()
        t.header.frame_id = 'shoulder0'
        t.child_frame_id = 'camera_link'
        t.transform = self.transform
        # t.transform.translation.x = 0.13395628663621322
        # t.transform.translation.y = 0.07813541504103579
        # t.transform.translation.z = 0.06372504768866594
        # t.transform.rotation.x = 0.0
        # t.transform.rotation.y = 0.0
        # t.transform.rotation.z = 0.0
        # t.transform.rotation.w = 1.0

        self.tf_broadcaster.sendTransform(t)


def main():
    rclpy.init()
    node = CameraFrameBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    rclpy.shutdown()

File: gh360/src/camera_frame.cpp
#include "gh360/camera_frame.hpp"

CameraFrame::CameraFrame() : Node("camera_frame")
{
    this->tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);


    this->transform = geometry_msgs::msg::TransformStamped();
    this->transform.header.stamp = this->now();
    this->transform.header.frame_id = "shoulder0";
    this->transform.child_frame_id = "camera_link";
    this->transform.transform.translation.x = 0.13395628663621322;
    this->transform.transform.translation.y = 0.07813541504103579;
    this->transform.transform.translation.z = 0.06372504768866594;
    this->transform.transform.rotation.x = 0.18302276459646596;
    this->transform.transform.rotation.y = 0.5245159060608517;
    this->transform.transform.rotation.z = -0.11628237572410241;
    this->transform.transform.rotation.w = 0.8233250518626218;

    this->timer_ = this->create_wall_timer(100ms, std::bind(&CameraFrame::timer_callback, this));

    RCLCPP_INFO(this->get_logger(), "Camera Frame node started");
}

CameraFrame::~CameraFrame()
{
}

void CameraFrame::timer_callback()
{
    this->transform.header.stamp = this->now();

    this->tf_broadcaster_->sendTransform(this->transform);
}

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);

    auto camera_frame_node = std::make_shared<CameraFrame>();
    rclcpp::spin(camera_frame_node);
    rclcpp::shutdown();

    return 0;
}